### PR TITLE
Added a listener to BoxAPIConnection

### PR DESCRIPTION
**Pull request by sarattall**
The BoxAPIConnection object now supports BoxAPIConnectionListeners which
enable the listener to listen to refresh event

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @sarattall has signed the CLA. Thanks for the pull request!

File: src/main/java/com/box/sdk/BoxAPIConnection.java [hunk 1]
```diff
@@ -283,7 +287,7 @@ public boolean needsRefresh() {
         boolean needsRefresh;
 
         this.refreshLock.readLock().lock();
-        if (this.expires == 0) {
+        if (this.expires < 0) {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: Does this change affect anything? I'd rather leave it out if it isn't needed.

**Comment by gcurtis** [comment]
Looks good! I'll merge as soon as the checkstyle errors are fixed.

File: src/main/java/com/box/sdk/BoxAPIConnection.java (hunk 1 above)
Comment (sarattall) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: We set expires to 0 to force the API connection to restart.

- -1 --> "null", "not set"
- 0  --> "force refresh the next time we need to"
- n  --> "needs refresh in n milliseconds"

Comment (sarattall) [inline src/main/java/com/box/sdk/BoxAPIConnection.java]: I see, your `-1` was our `0` and vice versa. Will revert this change
